src/lib/io/sol-uart-impl-riot.c

```diff
@@ -29,63 +29,120 @@
 #include "sol-mainloop.h"
 #include "sol-util-internal.h"
 #include "sol-interrupt_scheduler_riot.h"
+#include "sol-util.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
+#define DEFAULT_BUFFER_SIZE (128)
+
 struct sol_uart {
-    uart_t id;
-    struct {
-        void *handler;
-
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        struct sol_timeout *tx_writer;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length;
-    } async;
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, const struct sol_buffer *buf);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    const void *user_data;
+    struct sol_timeout *tx_writer;
+    struct sol_timeout *rx_reader;
+    void *handler;
+    struct sol_ptr_vector pending_blobs;
+    struct sol_buffer rx;
+    size_t pending_tx;
+    size_t tx_size;
+    int id;
+    bool delete_me;
+    bool in_use;
 };
 
+static bool
+rx_timeout(void *data)
+{
+    struct sol_uart *uart = data;
+    bool keep_running = true;
+    ssize_t r;
+
+    uart->in_use = true;
+    r = uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+    uart->in_use = false;
+    SOL_INT_CHECK(r, < 0, true);
+
+    if ((size_t)r < uart->rx.used) {
+        memmove(uart->rx.data, (char *)uart->rx.data + r, uart->rx.used - r);
+        uart->rx.used -= r;
+    } else if ((size_t)r >= uart->rx.used)
+        uart->rx.used = 0;
+
+    if (uart->delete_me)
+        sol_uart_close(uart);
+    else if (!uart->rx.used) {
+        uart->rx_reader = NULL;
+        keep_running = false;
+    }
+
+    return keep_running;
+}
+
 static void
 uart_rx_cb(void *arg, uint8_t data)
 {
     struct sol_uart *uart = arg;
+    size_t remaining;
+    int r;
 
-    if (!uart->async.rx_cb)
+    if (!uart->rx_cb)
         return;
-    uart->async.rx_cb((void *)uart->async.rx_user_data, uart,
-        (uint8_t)data);
+
+    remaining = uart->rx.capacity - uart->rx.used;
+    if (!remaining && !(uart->rx.flags & SOL_BUFFER_FLAGS_FIXED_CAPACITY)) {
+        r = sol_buffer_expand(&uart->rx, DEFAULT_BUFFER_SIZE);
```

Comment:
I'd not do this for small OS. We do this on Linux so we can read a full chunk at once. Here there is a single byte, then let the buffer grow policy to work on its own (multiples of 2 in the current implementation)